File: sdk/src/main/java/com/androidwind/androidquick/common/SafeHandler.java
package com.androidwind.androidquick.common;

import android.os.Handler;
import android.os.Message;

import java.lang.ref.WeakReference;

/**
 * @author ddnosh
 * @website http://blog.csdn.net/ddnosh
 */
public abstract class SafeHandler<T> extends Handler {

    private final WeakReference<T> mPage;

    public SafeHandler(T page) {
        mPage = new WeakReference<>(page);
    }

    @Override
    public void handleMessage(Message msg) {
        super.handleMessage(msg);
        T page = mPage.get();
        if (page != null) {
            disposeMessage(page, msg);
        }
    }

    public abstract void disposeMessage(T t, Message msg);
}

File: sdk/src/main/java/com/github/ddnosh/arabbit/util/handler/SafeHandler.kt
package com.github.ddnosh.arabbit.util.handler

import android.os.Handler
import android.os.Message

import java.lang.ref.WeakReference

/**
 * 防止内存泄漏的Handler
 * @author ddnosh
 * @website http://blog.csdn.net/ddnosh
 */
abstract class SafeHandler<T>(view: T) : Handler() {

    private var mView: WeakReference<T> = WeakReference(view)

    override fun handleMessage(msg: Message) {
        super.handleMessage(msg)
        val view = mView.get()
        view?.let {
            disposeMessage(it, msg)
        }
    }

    abstract fun disposeMessage(t: T, msg: Message)
}
